README.md
# listal-dl

[![Join the chat at https://gitter.im/vintol/listal](https://badges.gitter.im/vintol/listal.svg)](https://gitter.im/vintol/listal?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
batch download photos from listal.com

## [Listal.com](http://www.listal.com) - List the stuff you love!

![Listal-dl Poster](https://raw.githubusercontent.com/vintol/listal-dl/gh-pages/poster.png)


listal-dl can download all photos of any person from listal.com.

# Installation

see installation page [HERE](https://github.com/vintol/listal-dl/wiki/Installation).

# Download

Latest Release Page for [Downloads](https://github.com/vintol/listal-dl/releases/tag/0.21)

# Usage

see usage page [HERE](https://github.com/vintol/listal-dl/wiki/Uv0.2)

# Contact

mail me : vintol@mailinator.com


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

* * *

## [Version] - Release Date

> Added for new features.
> Changed for changes in existing functionality.
> Deprecated for soon-to-be removed features.
> Removed for now removed features.
> Fixed for any bug fixes.
> Security in case of vulnerabilities.

-------

## [1.2.2] - 2019-03-31

### Added

-   Cookie Support.

* * *

## [1.2.1] - 2018-01-27

### Added

-   Capability for handling https protocol.
    ### Deprecated
-   Forcing usage of http protocol because ssl certificate issues
    ### Security
-   HTTPS certificate is not verified.

* * *

## [1.2.0] - 2018-11-21

### Added

-   Stop scraping at a perticular image
-   Start from or finish at a perticular page no.
    ### Changed
-   Report Broken links.
    ### Fixed
-   Error handling exception when http timeout

* * *

## [1.1.0] - 2017-03-29

### Added

-   Downloader always checks if files are already downloaded and skips them.
-   Errors are categorized for better understanding of bugs.
-   Checks for broken links.

* * *

## [1.0.0] - 2016-11-10

First Stable release.

### Added

-   Can now get images from `image` lists.
-   Support for command-line Arguments.
-   Added a progress bar.
-   Download resume capability.

### Changed

-   Instead of entering the name of the person you enter the profile url.
-   Downloader is now a seperate script rather then an option.
-   Links are placed in a file in working dir; Instead of creating a seperate directory inside the working directory.
-   No interactive input; instead command line args are accepted.

### Removed

-   Option to download images is removed from main file.

* * *

## [0.1.0] - 2016-04-20

This is the first public release. **Pre Release**

### Added

-   Retrives the links of of all photos from any `person` profile on `listal.com`.
-   Also, can download the photos from the retrived links using mulitple threads.


L-dl.py
#
# Listal Downloader
# vintol.github.io/listal
#

import urllib.request,ssl
import argparse
import time
import queue
import threading
import os,sys
#

# Very Important. 
ssl._create_default_https_context = ssl._create_unverified_context

def download():
    global broken, failed, ls
    while not qq.empty():
        mydata = threading.local()
        mydata.name, mydata.url = qq.get()
        mydata.keep_going, mydata.skip, mydata.retry = (True, False, 0)
        #if mydata.name in ls:continue
        while mydata.keep_going:
            try:mydata.html = urllib.request.urlopen(mydata.url,timeout=120)
            except urllib.error.HTTPError as HERR:
                if HERR.code == 404 or HERR.code == 500:
                    broken += 1
                    mydata.keep_going = False
                    mydata.skip = True
            except:
                mydata.retry += 1
                if mydata.retry > 5:
                    mydata.keep_going = False
                    mydata.skip = True
            break
        if mydata.skip:continue
        while True:
            try:
                mydata.image = mydata.html.read()
                open(mydata.name,'wb').write(mydata.image)
                break
            except:
                mydata.retry += 1
                if mydata.retry > 10:break

def mkqueue():
    global total,ld,ls
    fhand = open(os.path.join(ld,args.fname),'r')
    links = []
    for each in fhand:
        if each.startswith('#') or len(each) < 10:continue
        fname = each.strip().split('/')[-2].zfill(10) + "." + each.strip().split('.')[-1]
        if fname not in ls:links.append((fname,each.strip()))
    for each in sorted(set(links),reverse=True):qq.put(each)
    total = qq.qsize()
    print(str(total),"Files queued for download.")
    fhand.close()

def enqueue():
    if qq.qsize() != 0:print("\n WARNING: Queue was not empty. ")
    for name,url in failed:
        qq.put((name,url))

def init_threads():
    for i in range(args.threads if args.threads < qq.qsize() else qq.qsize()):
        t = threading.Thread(target=download)
        threads.append(t)
        t.start()

def update_progress():
    progress = 100 - int((100*qq.qsize()) / total)
    te = time.strftime("%H:%M:%S",time.gmtime(time.time()-started))
    pbar = "\r {:0>3}% [{:<50}] ({},{}) Time Elapsed : {} ".format(progress, '#'*int((progress/2)), (total-qq.qsize()), total, te)
    sys.stdout.write(pbar)
    sys.stdout.flush()

def check_progress():
    global t1,p1
    t2 = time.time()
    p2 = 100 - int((100*qq.qsize()) / total)
    if t2-t1 < 25:pass
    else:
        if p2 - p1 >0:
            t1 = t2
            p1 = p2
        else:quit()
#

parser = argparse.ArgumentParser(description='Start Tao Downloader.')
parser.add_argument('fname', type=str,
                    help='The File containing list of links.')
parser.add_argument('--dir', dest='directory', type = str, default = None, required = False,
                    help='The directory to download files in.')
parser.add_argument('--threads', dest='threads', type = int, default = 10, required = False,
                    help='No. of threads to use.')
args = parser.parse_args()

#
qq = queue.Queue()
started = time.time()
threads = []
links   = []
failed  = []
broken  = 00
internal_error = 00
ld = os.getcwd()



if args.directory is not None:
    if not os.path.exists(args.directory):os.makedirs(args.directory)
    os.chdir(args.directory)
ls = os.listdir(os.getcwd())
#files = os.listdir(os.getcwd())

mkqueue()
init_threads()
t1, p1 = 0,0
while not qq.empty():
    update_progress()
    time.sleep(5)
    #check_progress()
for t in threads:t.join()

if len(failed) > 0:
    print("\n INFO : Download failed for {} Items. Trying Again ...".format(len(failed)))
    enqueue()
    failed.clear()
    init_threads()
    print("\n INFO:",len(failed),"Downloads Failed.")

print(" \n ============================ \n    Time Taken : {} \n Files Downloaded : {} \n    Failed Downloads\
    : {} \n    Broken Links : {}  \n ===============================================================".format(\
    time.strftime("%H:%M:%S",time.gmtime(time.time()-started)), total-len(failed),len(failed),broken))

#


listal-dl.py
#
# Listal-dl
#
# v0.21     28/08/2016
#
# Available under GNU GPL v3
#
#    listal-dl  Copyright (C) 2016  Tejas Kumar
#
#    This program comes with ABSOLUTELY NO WARRANTY.
#    This is free software, and you are welcome to redistribute it
#    under certain conditions; see file "LICENSE".
#
import urllib.request
from bs4 import *
import queue
import threading
import os
import time

##

class Imager (threading.Thread):
    def __init__(self, threadID, queue, lock, function, store):
        threading.Thread.__init__(self)
        self.threadID = threadID
        self.name = threadID
        self.queue = queue
        self.lock = lock
        self.execution_function = function
        self.output_store = store
    
    def run(self):
        while not self.queue.empty():
            self.lock.acquire()
            self.item = self.queue.get()
            print(self.name,"got item",self.item)
            self.lock.release()
            self.execution_function(self)
            if self.output_store is not None:
                self.lock.acquire()
                self.output_store.append(self.output)
                self.lock.release()
            self.queue.task_done()

#

def ipages(self):
    try:
        self.html = urllib.request.urlopen(self.item,timeout=2)
    except:
        while True:
            try:
                self.html = urllib.request.urlopen(self.item,timeout=5)
                if self.html.getcode() == 200:break
            except:continue
    try:self.html_data = self.html.read()
    except:
        self.lock.acquire()
        self.queue.put(self.item)
        self.lock.release()
        self.output = "\n"
        return
    self.soup = BeautifulSoup(self.html_data,"lxml")
    self.output = []
    for link in self.soup.find_all('a'):
        if link.get('href').startswith("http://www.listal.com/viewimage"):
            self.output.append(link.get('href')+"h")
#

def limages(self):
    try:
        self.html = urllib.request.urlopen(self.item,timeout=2)
    except:
        while True:
            try:
                self.html = urllib.request.urlopen(self.item,timeout=5)
                if self.html.getcode() == 200:break
            except:continue
    try:self.html_data = self.html.read()
    except:
        self.lock.acquire()
        self.queue.put(self.item)
        self.lock.release()
        self.output = "\n"
        return
    self.soup = BeautifulSoup(self.html_data,"lxml")
    self.output = self.soup.find(title=name).get('src')

#

def idownload(self):
    self.iname = self.item.split()[0]
    self.link = self.item.split()[1]
    try:
        self.html = urllib.request.urlopen(self.link,timeout=10)
    except:
        while True:
            try:
                self.html = urllib.request.urlopen(self.link,timeout=100)
                if self.html.getcode() == 200:break
            except:continue
    try:self.html_data = self.html.read()
    except:
        self.lock.acquire()
        self.queue.put(self.item)
        self.lock.release()
        return
    while True:
        try:
            open(self.iname,'wb').write(self.html_data)
        except:continue
        break

#

def pages():
    
    url_name = name.strip().lower().replace(' ','-')
    page_start = int(input("Start at Page No. : "))
    page_end = int(input("End at Page No. : ")) + 1
    no_threads = int(input("No. of Threads:"))
    
    for i in range(page_start,page_end):
        qq.put("http://www.listal.com/"+url_name+"/pictures//"+str(i))
     
    for n in range(no_threads):
        t = Imager("thread-{}".format(n),qq,thlock,ipages,output)
        threads.append(t)
        t.start()
    
    qq.join()
    
    for t in threads:
        t.join()
    
    for bulk in output:
        for link in bulk:
            links.append(link)

# Now Image Pages to Image Links
    
    for each in links:qq.put(each)
    output.clear()
    
    for n in range(no_threads):
        t = Imager("thread-{}".format(n),qq,thlock,limages,output)
        threads.append(t)
        t.start()
    
    qq.join()
    
    for t in threads:
        t.join()
    
    fhand = open("Images",'a')
    for link in output:
        fhand.write(link+"\n")

#

def images_download():
    
    links = open("Images",'r').read().split()
    if len(links) <= 8000:
        for i in range(len(links)):
            qq.put("{} {}".format("D"+str(1001+i)+".jpg",links[i]))
    elif len(links) > 8000:
        for i in range(8000):
            qq.put("{} {}".format("D"+str(1001+i)+".jpg",links[i]))
        for i in range(len(links)-8000):
            qq.put("{} {}".format("E"+str(1001+i)+".jpg",links[8000+i]))
    
    for n in range(int(input("No. of Threads:"))):
        t = Imager("thread-{}".format(n),qq,thlock,idownload,None)
        threads.append(t)
        t.start()
    
    qq.join()
    
    for t in threads:
        t.join()

##

print ("""    listal-dl  Copyright (C) 2016  Tejas Kumar

    This program comes with ABSOLUTELY NO WARRANTY.
    This is free software, and you are welcome to redistribute it
    under certain conditions; see file "LICENSE". \n """)

name = input("Name :")
qq = queue.Queue()
thlock = threading.Lock()
threads=[]
output=[]
links=[]

dir_name=name.split()[0]+name.split()[1][0]
dirs = os.listdir(os.getcwd())
if dir_name in dirs:
    print(dir_name,"already exists !")
    os.chdir(dir_name)
    print("Moving to directory :",os.getcwd())
else:
    os.mkdir(dir_name)
    os.chdir(dir_name)
    print("Moving to directory :",os.getcwd())


choise = input (" 0] Get Image Links \n 1] Download Images \n ===> ")
time_started = time.time()
if choise == "0":pages()
elif choise == "1":images_download()
else: print("Try Again.")

time_taken = time.time() - time_started
print("Time Taken = {}:{}:{}".format(str(int(time_taken/3600)).zfill(2),str(int((time_taken%3600)/60)).zfill(2),str(int((time_taken%3600)%60)).zfill(2)))

##

Listal.py
#   Listal.py
#   08/11/2016 - 31/03/2019
#   v 1.2.2

import urllib.request, urllib.parse
import http.cookiejar, ssl
import bs4
import queue
import threading
import re
import os
import sys
import argparse
import time

# Scrapers

def get_ipages():
    global IMG, STOP_AT
    while not qq.empty():
        local = threading.local()
        local.url = qq.get()
        local.keep_going = True
        local.skip = False
        if STOP_AT is not None and int(local.url.split('//')[2]) > STOP_AT:continue
        while local.keep_going:
            try:local.html = urllib.request.urlopen(local.url,timeout=10)
            except urllib.error.HTTPError as HERR:
                if HERR.code == 404:
                    local.keep_going = False
                    local.skip = True
                    continue
            except:continue
            if local.html.getcode() == 200:local.keep_going = False
        if local.skip:continue
        local.data = local.html.read()
        local.soup = bs4.BeautifulSoup(local.data,'lxml')
        for each in local.soup.find_all('div','imagewrap-inner'):
            local.img = int(each.a.get('href').strip().split('/')[-1])
            if IMG is None:ipages.append(local.img)
            elif local.img > IMG:ipages.append(local.img)
            elif local.img == IMG:STOP_AT = int(local.url.split('//')[2])
            else:pass

def get_images():
    while not qq.empty():
        local = threading.local()
        local.url = qq.get()
        local.keep_going = True
        local.skip = True
        local.retry = 0
        while local.keep_going and local.retry < 5:
            try:
                local.retry += 1
                local.html = urllib.request.urlopen(local.url,timeout=25)
                if local.html.getcode() == 200:
                    local.keep_going = False
                    local.skip = False
            except urllib.error.HTTPError as HERR:
                if HERR is not None and HERR.code == 404:
                    local.keep_going = False
                    continue
            except:continue
        if local.skip:continue
        for i in range(2):
            try:
                local.data = local.html.read()
                images.append(find_image(local.data))
            except:continue
            break

# Functions

def mksoup(url):
    tmp = urllib.request.urlopen(url)
    return bs4.BeautifulSoup(tmp.read(),"lxml")

def find_image(data):
    return bs4.BeautifulSoup(data,"lxml").find('img','pure-img').get('src').replace("https:","http:")

def post_req():
    tmp = urllib.parse.urlencode({ 'listid' : list_id , 'offset' : offset})
    return urllib.request.urlopen("https://www.listal.com/item-list/",tmp.encode())

def mkqueue(url):
    global no_pics,no_pages
    no_pics  = int(mksoup(url).find('a','picturesbutton').span.text.strip())
    no_pages = no_pics/50
    if no_pages.is_integer():no_pages = int(no_pages)
    else:no_pages = int(no_pages) + 1
    for i in range(int(args.first_page),no_pages+1):qq.put(url+"/pictures//"+str(i))

def enqueue():
    global qq,ipages
    if not qq.empty():print("WARNING : Queue was not empty.")
    qq = queue.Queue()
    ipages = sorted(set(ipages))
    for each in ipages:
        qq.put("http://www.listal.com/viewimage/"+str(each)+"h")

def stop_at(IMG):
    tmp = []
    for each in ipages:
        if each > IMG:tmp.append(each)
    ipages = tmp

def update_progress():
    progress = 100 - int((100*qq.qsize()) / len(ipages))
    pbar = "\r {:0>3}% [{:<50}] ({},{}) ".format(progress, '#'*int((progress/2)), (len(ipages)-qq.qsize()), len(ipages))
    sys.stdout.write(pbar)
    sys.stdout.flush()

def get_listinfo(url):
    global list_type,list_id,list_name,total_pic,offset
    soup = mksoup(url)
    list_type = soup.find(id='customlistitems').get('data-listformat')
    if list_type != "images":
        print("This is not a Image list. Currently listal.dl suppots only Image lists.")
        quit()
    list_id   = int(soup.find(id='customlistitems').get('data-listid'))
    try:list_name = soup.find('div','headertitle').text.strip()
    except AttributeError:list_name = urls.path[6:].replace('-',' ').title()
    total_pic = int(soup.find(id='customlistitems').div.get('data-itemtotal'))
    offset    = int(soup.find('div','loadmoreitems').get('data-offset'))
    for each in soup.find_all('div','imagelistbox'):
        ipages.append(int(each.a.get('href').strip().split('/')[-1]))

def get_list():
    global offset
    while True:
        data = post_req().read()
        for each in sorted(set(re.findall("viewimage\\\/([0-9]{4,10})'" ,data.decode()))):
            ipages.append(int(each))
            offset = offset + 1
        if offset == total_pic:break

def write():
    if urls.path.startswith("/list/"):fhand = open(list_name+".txt",'a')
    else:fhand = open(name+".txt",'a')
    fhand.write("### {} : {} Images\n".format(finished,len(images)))
    for each in images:fhand.write(each+"\n")
    fhand.close()

# Global

qq = queue.Queue()
threads = []
ipages  = []
images  = []
IMG     = None
STOP_AT = None
started = time.time()

# Main

parser = argparse.ArgumentParser(description='Scrape Images from \'listal.com\'.')
parser.add_argument('url', type=str,
                    help='URL to the List or Profile on listal.com.')
parser.add_argument('--from', dest='first_page', type = int, default = None, required = False,
                    help='The profile page no to start scraping images from')
parser.add_argument('--upto', dest='last_page' , type = int, default = None, required = False,
                    help='Scrap images only upto the page no.')
parser.add_argument('--threads', dest='threads', type = int, default = 10, required = False,
                    help='No. of threads to use.')
args = parser.parse_args()

urls = urllib.parse.urlparse(args.url)
if urls.netloc != 'www.listal.com':
    print ("Check the Entered URL.")
    quit()

#CookieJar Initiation
urllib.request.HTTPCookieProcessor(http.cookiejar.CookieJar())

if urls.path.startswith("/list/"):
    if args.first_page is not None:print("Entered URL is of a list. The '--from' option is ignored.")
    if args.last_page  is not None:print("Entered URL is of a list. The '--upto' option is ignored.")
    get_listinfo(urls.geturl())
    get_list()
else:
    urls = urllib.parse.urlparse(urls.geturl().split('/picture')[0])
    name = urls.path[1:].replace('-',' ').title()
    if args.first_page is None:args.first_page = 1
    if args.last_page is not None:
        for i in range(args.first_page,args.last_page+1):qq.put(args.url+"/pictures//"+str(i))
    else:mkqueue(urls.geturl())
    for n in range(args.threads):
        t = threading.Thread(target=get_ipages)
        threads.append(t)
        t.start()
    for t in threads:t.join()

print("Phase I Complete.",len(ipages),"Images Found.")
print("Time Taken :",time.strftime("%H:%M:%S",time.gmtime(time.time()-started)))
print("Phase II :")
enqueue()
threads.clear()
for n in range(args.threads):
    t = threading.Thread(target=get_images)
    threads.append(t)
    t.start()

while not qq.empty():
    update_progress()
    sys.stdout.flush()
    time.sleep(1)

for t in threads:t.join()

time_taken = time.time() - started
finished   = time.strftime("%d/%m/%Y %H:%M",time.localtime())
write()
print("Time Taken :",time.strftime("%H:%M:%S",time.gmtime(time_taken)))

# END
